cpp/src/replace/replace.cu

```diff
@@ -13,26 +16,551 @@
+    if(threadIdx.x < warp_size) valid_sum[threadIdx.x] = 0;
+    __syncthreads();
+  }
+
+  while (i < nrows) {
+    bool output_is_valid = true;
+    uint32_t bitmask = 0xffffffff;
+
+    if (input_has_nulls) {
+      bool const input_is_valid{cudf::bit_is_set(input_valid, i)};
+      output_is_valid = input_is_valid;
+
+      bitmask = __ballot_sync(active_mask, input_is_valid);
+
+      if (input_is_valid) {
+        thrust::tie(output_data[i], output_is_valid)  =
+            get_new_value<T, replacement_has_nulls>(i, input_data,
+                                      values_to_replace_begin,
+                                      values_to_replace_end,
+                                      d_replacement_values,
+                                      replacement_valid);
+      }
+
+    } else {
+       thrust::tie(output_data[i], output_is_valid) =
+            get_new_value<T, replacement_has_nulls>(i, input_data,
+                                      values_to_replace_begin,
+                                      values_to_replace_end,
+                                      d_replacement_values,
+                                      replacement_valid);
+    }
+
+    /* output valid counts calculations*/
+    if (input_has_nulls or replacement_has_nulls){
+
+      bitmask &= __ballot_sync(active_mask, output_is_valid);
+
+      if(0 == (threadIdx.x % warp_size)){
+        output_valid[(int)(i/warp_size)] = bitmask;
+        valid_sum[(int)(threadIdx.x / warp_size)] += __popc(bitmask);
+      }
+    }
+
+    i += blockDim.x * gridDim.x;
+    active_mask = __ballot_sync(active_mask, i < nrows);
+  }
+  if(input_has_nulls or replacement_has_nulls){
+    __syncthreads(); // waiting for the valid counts of each warp to be ready
+
+    // Compute total valid count for this block and add it to global count
+    uint32_t block_valid_count = sum_warps<uint32_t>(valid_sum);
+
+    // one thread computes and adds to output_valid_count
+    if (threadIdx.x < warp_size && 0 == (threadIdx.x % warp_size)) {
+      atomicAdd(output_valid_count, block_valid_count);
+    }
+  }
+}
+
+  /* --------------------------------------------------------------------------*/
+  /**
+   * @brief Functor called by the `type_dispatcher` in order to invoke and instantiate
+   *        `replace_kernel` with the appropriate data types.
+   */
+  /* ----------------------------------------------------------------------------*/
+  struct replace_kernel_forwarder {
+    template <typename col_type>
+    void operator()(cudf::column_view const& input_col,
+                    cudf::column_view const& values_to_replace,
+                    cudf::column_view const& replacement_values,
+                    cudf::mutable_column_view& output,
+                    cudaStream_t stream = 0,
+                    rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource())
+    {
+      const bool input_has_nulls = input_col.has_nulls();
```

Comment:
These look redundant. Just call the functions where necessary. 
```suggestion
```

Reply:
Done.